Try to improve the accessibility of featured images

jasmussen opened the pull request:
The goal of this PR is to fix #1116. It replaces #3829 as this is better code, and the rebase was also too complex. 

This PR consists of two commits. 

The first commit, https://github.com/WordPress/gutenberg/commit/76f19466b6e12e7ad7ab2cbaedb5e5e2ea1e7e97, is mostly a copy of the old PR (#3829), but relies on the exact same labels that we receive from `postLabel`. 

The 2nd commit, https://github.com/WordPress/gutenberg/commit/6a408274ff4a5171e4abb4dbb284833732a27c86, removes the reliance on the labels we receive frmo `postLabel` entirely, because they do not match what we are doing with multiple buttons, and in my understanding, those separate and multiple buttons are key to improving accessibility.

I would like comments on both approaches, and the approach of this PR overall. Here's a GIF:

![featured](https://user-images.githubusercontent.com/1204802/41090728-38c21af6-6a45-11e8-91cf-38745ad96a5f.gif)

Commit:
Try improving the a11n of the featured image feature

This commit reproduces #3829, but with slightly better code. The rebase was also too complex.

Commit:
Try customizing the labels instead of relying on those from the postLabel

youknowriad commented on editor/components/post-featured-image/index.js:
```diff
@@ -50,26 +46,47 @@ function PostFeaturedImage( { featuredImageId, onUpdateImage, onRemoveImage, med
-					/>
+					<div>
+						<MediaUpload
+							title={ __( 'No image selected' ) }
```
This title prop is the title of the modal I think, Are you certain you want this? Shoudl this also be `Add image`?

youknowriad reviewed (commented):
Do you think we should show the buttons (replace, remove) on the same line design-wise?

Commit:
Fix modal title, and let blocks align next to each other

jasmussen commented:
Great feedback. No, didn't want to change the modal title, and letting the buttons float is cool too:

<img width="274" alt="screen shot 2018-06-07 at 12 20 52" src="https://user-images.githubusercontent.com/1204802/41094119-6dc7b1fe-6a4d-11e8-96ef-ef8e387b8d11.png">

youknowriad commented on editor/components/post-featured-image/index.js:
```diff
@@ -50,26 +46,47 @@ function PostFeaturedImage( { featuredImageId, onUpdateImage, onRemoveImage, med
+							) }
+						/>
+						<MediaUpload
+							title={ __( 'Add Image' ) }
```
I think we should use the same title in the modal here as well "Set featured image'?

afercia commented:
@jasmussen thanks for this PR 🙂One of the things I was trying to explain on #1116 is that there's no need for two buttons that do the same thing:

<img width="300" alt="screen shot 2018-06-07 at 18 07 28" src="https://user-images.githubusercontent.com/1682452/41112161-804b99f4-6a7e-11e8-8f9b-2e908aaa2b55.png">

- the "plaecholder" is a button, when clicked opens the media modal
- same for the "Add Image" button

To me, seems redundant and potentially confusing for assistive technology users. I'd propose to further simplify. I'd say the "state" ("No image selected") is pretty evident because there's no image... do we really need a text to communicate the state? Also, the button says "No image selected" so it doesn't explicitly communicate the underlying action.

Removing the second button and changing the placeholder-button text could work perhaps, and would further simplify the code:

<img width="304" alt="screen shot 2018-06-07 at 18 08 32" src="https://user-images.githubusercontent.com/1682452/41112366-03421874-6a7f-11e8-9f9b-7903064c89cd.png">

Thoughts?

Lastly, when an image has been set, the image itself is within a button so it's using a control with native keyboard interaction. That's good. We just need to use a smart `alt` text on the image. That would work as the accessible name of the button-image. I think @anevins12 is willing to try to address that in a separate PR. 🙂

afercia reviewed (commented):
See my last comment!

anevins12 commented:
I'm happy to address the alt text in a separate PR

melchoyce commented:
I think this is a good solution:

<img src="https://user-images.githubusercontent.com/1682452/41112366-03421874-6a7f-11e8-9f9b-7903064c89cd.png">

karmatosed commented:
I would second what has been said here, I don't think we need to have buttons here and would love to see that explored.

jasmussen commented:
🤘

I'll update this tomorrow to reflect feedback.

Commit:
Address feedback.

Commit:
Fix modal title.

jasmussen commented:
I think I addressed the feedback now, both with the modal title and the removal of the additional button.

![featured](https://user-images.githubusercontent.com/1204802/41147339-6f93d8dc-6b06-11e8-9a40-ab4fdcd5472f.gif)

What's still missing is the alt title for the image — @anevins12 if you'd like to address that separately, that would be cool, thanks!

youknowriad commented on editor/components/post-featured-image/index.js:
```diff
@@ -30,12 +26,12 @@ function PostFeaturedImage( { featuredImageId, onUpdateImage, onRemoveImage, med
 				{ !! featuredImageId &&
 					<MediaUpload
-						title={ postLabel.set_featured_image }
+						title={ __( 'Set featured image' ) }
```
Any reason to avoid the labels from the post object `postLabel.set_featured_image` and `postLabel.remove_featured_image`. These labels are tweakable server-side. We could use the proposed labels as a fallback if the `set_featured_image` or `remove_featured_image` are not defined?

jasmussen replied:
I've no strong opinion here. In this case I changed it so it was consistent with the other labels.

"Set featured image" is the only label we haven't customized. If you'd like, I can restore all instances where we use "Set featured image" as the label to use the postLabel again.

youknowriad replied:
I feel we should restore it essentially for Backwards Compatibility because CPT are probably customizing this label. And I feel we consistently use CPT's labels across the app. In which components are we avoiding those?

jasmussen replied:
I have no moral objections to reapplying those labels. But here's what's in master:

```
const DEFAULT_SET_FEATURE_IMAGE_LABEL = __( 'Set featured image' );
const DEFAULT_REMOVE_FEATURE_IMAGE_LABEL = __( 'Remove featured image' );
```

Those labels match the ones from `postLabel`.

If this branch gets merged sans changes, then we will have the following labels:

- `__( 'Set featured image' )` — this could match the postLabel
- `__( 'Replace Image' )` — no match in postLabel
- `__( 'Remove Image' )` — could match "Remove featured image", but it feels super redundant and would take up more space.

The postLabels are definitely designed for the old interface:

<img width="306" alt="screen shot 2018-06-08 at 10 47 11" src="https://user-images.githubusercontent.com/1204802/41148554-86ae441e-6b09-11e8-9b68-4fe7e88f3045.png">

If we can update the postLabel labels, that would be ideal, but I imagine that's not possible. 

What do you think we should do?

youknowriad replied:
I personally don't think the labels for the featured images are that important because it's a generic thing:

 - We can try to add the `replace_featured_image` label to the CPT labels. (Not certain how we define the existence of these labels)
 - I'm also fine considering the featured image generic enough to avoid these labels but maybe we can ask someone more familiar with the plugins to know whether these are heavily used or not? I don't feel confident making the decision myself.

jasmussen replied:
Solid thoughts.

@pento, any idea here?

jasmussen replied:
@afercia you also have a treasure trove of core experience — do you have any insights here?

afercia replied:
@jasmussen sorry, I'm late to the party. Is your question related to how many plugins might use custom labels? Not a great plugins expert here, but I guess it's pretty common for plugins to customize them, just a couple examples: EDD and Woocommerce:
```
Easy digital downloads:
'featured_image'        => __( '%1$s Image', 'easy-digital-downloads' ),
'set_featured_image'    => __( 'Set %1$s Image', 'easy-digital-downloads' ),
'remove_featured_image' => __( 'Remove %1$s Image', 'easy-digital-downloads' ),
'use_featured_image'    => __( 'Use as %1$s Image', 'easy-digital-downloads' ),

Woocommerce:
'featured_image'        => __( 'Product image', 'woocommerce' ),
'set_featured_image'    => __( 'Set product image', 'woocommerce' ),
'remove_featured_image' => __( 'Remove product image', 'woocommerce' ),
'use_featured_image'    => __( 'Use as product image', 'woocommerce' ),
```

karmatosed reviewed (approved):
Looks great to me from design view.

paulwilde commented:
Separate PR, but this new UI is primed for also including drag-and-drop support!

tofumatt commented on editor/components/post-featured-image/index.js:
```diff
@@ -50,26 +46,36 @@ function PostFeaturedImage( { featuredImageId, onUpdateImage, onRemoveImage, med
-						) }
-					/>
+					<div>
+						<MediaUpload
```
This looks like the exact same component/args/markup as above. Any chance we could refactor this a bit to reduce the duplication?

tofumatt commented on editor/components/post-featured-image/style.scss:
```diff
@@ -1,33 +1,38 @@
-		outline: none;
+	// Space consecutive buttons evenly.
+	.components-button + .components-button {
+		margin-top: 1em;
```
Eventually I'll propose sorting our CSS attributes alphabetically; until then I'll live with this 😉

jasmussen replied:
:) I like your style.

Though elaborate a bit — are you referring to the top margin going before the right margin? For north south east west things I always go clockwise, i.e. top, right, bottom, left.

tofumatt commented on editor/components/post-featured-image/style.scss:
```diff
@@ -1,33 +1,38 @@
 
-	&:hover {
-		color: theme( primary );
+	// Don't size up images beyond their intrinsic size.
```
"size up" is a bit odd, maybe:

```css
// This keeps images at their intrinsic size (eg. a 50px
// image will never be wider than 50px).
```

tofumatt reviewed (approved):
> What's still missing is the alt title for the image — @anevins12 if you'd like to address that separately, that would be cool, thanks!

Code looks good but merge after creating a follow-up issue for that, please 😄

Commit:
Address feedback, restore postLabel.

tofumatt reviewed (commented):
I hate to nitpick but the left-aligned buttons under each other, both starting with "Re" in English, I find quite hard to tell apart. If it should be a new/separate UX issue that's cool but "Replace"/"Remove" are so similar in length/look and the buttons have the same style, so I find those buttons very hard to scan quickly.

If we wanna do that as a follow-up though, that's cool. Let me know and I'll file it.

jasmussen commented:
I could change it to say "Replace featured image", then the two buttons would be closer to each other in length. But that would be arguably worse with regards to the similarity between the buttons. What do you think?

I personally really liked the old "Replace image" "Remove image" buttons, but the gnarly reasons for why this isn't smart start here: https://github.com/WordPress/gutenberg/pull/7200#discussion_r193987588 — back compat, in other words :(

tofumatt commented:
I think they could use different styles (maybe a more “scary”/red remove button). The issue is they are similar lengths with similar-looking wording in English on top of each other with no style distinctions. As long as they both start with “Re” I think it’s a problem.

> On 20 Jun 2018, at 09:38, Joen Asmussen <notifications@github.com> wrote:
> 
> I could change it to say "Replace featured image", then the two buttons would be closer to each other in length. But that would be arguably worse with regards to the similarity between the buttons. What do you think?
> 
> I personally really liked the old "Replace image" "Remove image" buttons, but the gnarly reasons for why this isn't smart start here: #7200 (comment) <https://github.com/WordPress/gutenberg/pull/7200#discussion_r193987588> — back compat, in other words :(
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/WordPress/gutenberg/pull/7200#issuecomment-398654470>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAFi99eiuHynkcHfWrPfS8c1v9kPJCrGks5t-fwPgaJpZM4UeC3I>.
>

jasmussen commented:
Here's what we do on wp.com:

<img width="432" alt="screen shot 2018-06-20 at 09 50 10" src="https://user-images.githubusercontent.com/1204802/41644765-5994cc36-746f-11e8-8271-230672a0d659.png">

We could make it a positioned `IconButton` with an X, and have "Remove image" be the tooltip?

tofumatt commented:
OMG that's so much better. 👍

Commit:
Tweak "Remove" button.

jasmussen commented:
Tweaked:

![now](https://user-images.githubusercontent.com/1204802/41646189-6aa17cf0-7473-11e8-953d-56166d7adebb.gif)

This is a biggish change, so CC: @afercia @karmatosed for re-review.

I also added a focus style to the image itself, the focus style was hidden underneath the image itself.

tofumatt commented:
Oops, darn. Would adding "Remove" or changing the "x" to "remove" work? It could be "Remove image" with "image" as off-screen text for screenreader.

As a sighted user, having the different actions in different places and looking different __vastly__ improves usability.

jasmussen commented:
> If the problem is in the button text, than the we could also consider to update the labels used by WordPress.

That is the primary problem.

But those are from the `postLabel`, and unless those are easy to change, I need a plan for how to proceed and get this PR merged.

karmatosed commented:
In this case let's take a step back have 2 buttons, but then have another issue to iterate - I think holding it up for that discussion isn't the best option.

@afercia whilst I understand the accessibility concerns and I am recommending we go back to 2 buttons temporarily, I would like to find an option that is both accessible and usable. Having double buttons looking the same simply isn't usable for a lot of people. Are there any examples you can think of that achieve this we can learn from?

Commit:
Revert last button.

Commit:
Clean up CSS.

jasmussen commented:
Okay, we're back to this:

<img width="282" alt="screen shot 2018-06-20 at 12 27 20" src="https://user-images.githubusercontent.com/1204802/41653115-5a9df7cc-7485-11e8-884f-526afea88730.png">

I think it might be good to merge this in and iterate separately if need be, including adjusting the `postLabel`s if need be. 

We can also go crazy and add a "scary" button color:

<img width="350" alt="screen shot 2018-06-20 at 12 29 50" src="https://user-images.githubusercontent.com/1204802/41653259-d05c3f28-7485-11e8-8e55-f696b3d26c68.png">

But that also seems overkill to me.

tofumatt commented:
Or you could make the "replace image” button blue or something if the red is too much. Just *anything* to distinguish them visually.

jasmussen commented:
The blue button is supposed to be a "primary" button, and I feel like UX-wise there is no primary action here. Both actions are secondary, in that you're basically "done" with what you set out to do, which is to set a featured image.

tofumatt commented:
Good point, I worried about that.

I am for red then, but I am not a designer so feel free to ignore. It looks weird but it is more usable to me.

afercia commented:
> Are there any examples you can think of that achieve this we can learn from?

@karmatosed hm so this is especially related to the "remove" button sitting close to the replace one, correct? I think the most recent design discussions and implementations about placement of multiple buttons and about how to distinguish the one used for destructive action were for the Media widgets and the new pattern recently implemented in the tags/categories edit screens.

The media widgets have multiple button, but the control for destructive action is completely separated:

![screen shot 2018-06-20 at 12 33 16](https://user-images.githubusercontent.com/1682452/41653678-213a571c-7487-11e8-86be-b3471336d800.png)

The pattern used for tags/categories is more interesting, perhaps. If I recall correctly, it was discussed at length from a design perspective:

![screen shot 2018-06-20 at 12 35 20](https://user-images.githubusercontent.com/1682452/41653739-4efd6d2e-7487-11e8-9456-e59f306f920c.png)

I'm not saying to make the "replace" image blue, but maybe differentiating the "remove" one as @jasmussen suggested could be a good option?

@jasmussen don't be afraid to open a ticket on Trac to change the labels used by WordPress 🙂if that makes things more usable and adds value, I guess everybody would be happy to change them.

tofumatt commented:
That red, text-style button would work for me... I know there are also a11y/UX concerns with buttons looking like links/not like buttons, but if "Replace" was a standard button and "Remove image" was a red-text/link-styled button I would find this more usable... and I think it's prettier than the scary red button. Thanks @afercia! 👍

jasmussen committed:
Add isScary prop to button.

jasmussen commented:
Okay, now it's this:

![screen shot 2018-06-20 at 13 33 03](https://user-images.githubusercontent.com/1204802/41655870-8a9a17fe-748e-11e8-9c8d-774ddb36d095.png)

@tofumatt if you could look at the extra prop:https://github.com/WordPress/gutenberg/pull/7200/commits/20f54898e955cb9c9607282ce34a3125bf37e4dd — this felt the sensible way to do it for me?

jasmussen commented:
> don't be afraid to open a ticket on Trac to change the labels used by WordPress 🙂if that makes things more usable and adds value, I guess everybody would be happy to change them.

I'm not at all afraid of doing that, and we should certainly do this regardless. But we can move faster with Gutenberg than we can with the rest of core, which means it should be staggered — first this, then that.

tofumatt reviewed (commented):
I don't see "scary" as a prop or style elsewhere in the code. Is it in a style guide or something?

I'd maybe go with `isDestructive` over `isScary`.

jasmussen commented:
Yes, there's a "scary" term used in a few style guides I've seen, including the wp.com one. There's nothing for WordPress, the core css class is just "delete". I can go with destructive.

jasmussen committed:
isScary to isDestructive

jasmussen commented:
Okay, it's now destructive instead of scary :)

tofumatt commented:
I dig it 👍🏻
